File: Day #455/solution.py
# Day 455: Conway's Game of Life on an infinite board.
# Set of live cells; tally neighbours each tick. Time O(live) per step.
from collections import Counter


def step(live):
    cnt = Counter()
    for (r, c) in live:
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr or dc:
                    cnt[(r + dr, c + dc)] += 1
    nxt = set()
    for cell, n in cnt.items():
        if n == 3 or (n == 2 and cell in live):
            nxt.add(cell)
    return nxt


def print_board(live):
    if not live:
        print("(empty)")
        return
    rs = [r for r, _ in live]
    cs = [c for _, c in live]
    for r in range(min(rs), max(rs) + 1):
        print("".join('*' if (r, c) in live else '.'
                       for c in range(min(cs), max(cs) + 1)))


if __name__ == "__main__":
    live = {(1, 0), (1, 1), (1, 2)}  # horizontal blinker
    steps = 2
    print("Step 0:")
    print_board(live)
    for s in range(1, steps + 1):
        live = step(live)
        print(f"Step {s}:")
        print_board(live)

File: Day #455/Solution.java
// Day 455: Conway's Game of Life on an infinite board.
// Hash set of live cells; tally neighbours each tick. Time O(live) per step.
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Solution {
    static long key(int r, int c) { return (((long) r) << 32) ^ (c & 0xffffffffL); }
    static int row(long k) { return (int) (k >> 32); }
    static int col(long k) { return (int) k; }

    static Set<Long> step(Set<Long> live) {
        Map<Long, Integer> cnt = new HashMap<>();
        for (long c : live) {
            int r = row(c), cc = col(c);
            for (int dr = -1; dr <= 1; dr++)
                for (int dc = -1; dc <= 1; dc++)
                    if (dr != 0 || dc != 0)
                        cnt.merge(key(r + dr, cc + dc), 1, Integer::sum);
        }
        Set<Long> next = new HashSet<>();
        for (Map.Entry<Long, Integer> e : cnt.entrySet()) {
            int n = e.getValue();
            boolean alive = live.contains(e.getKey());
            if (n == 3 || (alive && n == 2)) next.add(e.getKey());
        }
        return next;
    }

    static void printBoard(Set<Long> live) {
        if (live.isEmpty()) { System.out.println("(empty)"); return; }
        int r0 = Integer.MAX_VALUE, r1 = Integer.MIN_VALUE, c0 = Integer.MAX_VALUE, c1 = Integer.MIN_VALUE;
        for (long k : live) {
            r0 = Math.min(r0, row(k)); r1 = Math.max(r1, row(k));
            c0 = Math.min(c0, col(k)); c1 = Math.max(c1, col(k));
        }
        StringBuilder sb = new StringBuilder();
        for (int r = r0; r <= r1; r++) {
            for (int c = c0; c <= c1; c++)
                sb.append(live.contains(key(r, c)) ? '*' : '.');
            sb.append('\n');
        }
        System.out.print(sb);
    }

    public static void main(String[] args) {
        Set<Long> live = new HashSet<>();
        live.add(key(1, 0)); live.add(key(1, 1)); live.add(key(1, 2));
        int steps = 2;
        System.out.println("Step 0:"); printBoard(live);
        for (int s = 1; s <= steps; s++) {
            live = step(live);
            System.out.println("Step " + s + ":");
            printBoard(live);
        }
    }
}
